components/appmgt-extensions/org.wso2.carbon.appmgt.mdm.wso2mdm.direct/src/main/java/org/wso2/carbon/appmgt/mdm/wso2mdm/direct/MDMOperationsImpl.java

```diff
@@ -0,0 +1,308 @@
+            return Collections.emptyList();
+        }
+
+        List<Device> result = new ArrayList<>();
+        try {
+            List<org.wso2.carbon.device.mgt.common.Device> mdmDevices = null;
+            switch (type) {
+
+            case APPM_APP_CALL_TYPE_USER:
+                if (params.length > 0) {
+                    String userName = params[0];
+                    mdmDevices = deviceManagementProviderService.getDevicesOfUser(userName);
+                } else {
+                    log.error("User name is not found in parameter [0]");
+                }
+                break;
+            case APPM_APP_CALL_TYPE_ROLE:
+                if (params.length > 0) {
+                    String roleName = params[0];
+                    mdmDevices = deviceManagementProviderService.getAllDevicesOfRole(roleName);
+                } else {
+                    log.error("Role name is not found in parameter [0]");
+                }
+                break;
+            case APPM_APP_CALL_TYPE_STATUS:
+                //                mdmDevices = deviceManagementProviderService.getDevicesByStatus(currentUser);
+                log.error("getDevices with type [status] is not supported on this version");
+                break;
+            default:
+                mdmDevices = deviceManagementProviderService.getAllDevices();
+            }
+
+            for (org.wso2.carbon.device.mgt.common.Device mdmDevice : mdmDevices) {
+                Device device = convertDevice(mdmDevice, configParams);
+                result.add(device);
+            }
+        } catch (DeviceManagementException e) {
+            log.error("Error occurred getting the device list for user " + currentUser);
+        }
+        return result;
+    }
+
+    private Device convertDevice(org.wso2.carbon.device.mgt.common.Device mdmDevice,
+            HashMap<String, String> configParams) {
+        Device device = new Device();
+        Map<String, String> props = new HashMap<>();
+        for (org.wso2.carbon.device.mgt.common.Device.Property property : mdmDevice.getProperties()) {
+            props.put(property.getName(), property.getValue());
+        }
+        device.setId(mdmDevice.getDeviceIdentifier() + APPM_DEVICE_ID_CONCAT_SEPARATOR + mdmDevice.getType());
+        device.setName(props.get(MDMAppConstants.MDMAppPropertyKeys.DEVICE_NAME));
+        device.setModel(props.get(MDMAppConstants.MDMAppPropertyKeys.DEVICE_MODEL));
+        device.setPlatform(mdmDevice.getType());
+        device.setImage(String.format(configParams.get(APPM_APP_IMAGE_URL), device.getModel()));
+        device.setType(APPM_APP_TYPE_MOBILE_DEVICE);
+        device.setPlatformVersion(props.get(MDMAppConstants.MDMAppPropertyKeys.OS_VERSION));
+
+        return device;
+    }
+
+    private DeviceManagementProviderService getDeviceManagerService(String tenantDomain) throws AppManagementException {
+        try {
+            PrivilegedCarbonContext.startTenantFlow();
```

Comment:
This I would say is a security vulnerability. Only the super tenant should be able to start a tenant flow on behalf of another tenant. So, the logic should first check if the currently logged in user is a super tenant and then proceed with the remaining bits.